### PR TITLE
Adding 5 tests that were failing on me: [ wtq-00080, wtq-00085, wtq-00113, wtq-00162, wtq-00168]

### PR DESCRIPTION
**Pull request by P6rguVyrst**
## Summary

Three round-trip bugs fixed, discovered via accuracy testing with WTQ benchmark datasets.

### 1. Boolean-like dictionary keys converted to booleans

**Problem:** `{"f": 1}` decoded as `{False: 1}`.

**Fix:** Encoder now quotes keys matching boolean/null keywords (t, f, true, false, null, none, nil). Decoder uses a new `parse_key()` function that preserves keys as strings.

### 2. Dictionary header values not quoted

**Problem:** Values containing commas, colons, or special characters were truncated at delimiters.

**Fix:** Uses `_format_value()` for consistent quoting of dictionary compression header values.

### 3. Float columns corrupted by delta encoding

**Problem:** Delta encoding violated spec §2.3's round-trip requirement. IEEE-754 arithmetic `prev + (cur - prev)` doesn't recover original bit patterns, and `round(diff, 10)` compounded the loss.

**Example:** `1865.43` decoded as `1865.4299999999994`

**Fix:** Restrict `SparseMode.DELTA` to int-only columns. Float columns use standard value encoding which round-trips exactly via Python's shortest-round-trip `str(float)`.

Regression tests cover: benchmark values, math.pi/math.e, 0.1+0.2, extreme exponents, negatives.

## Testing Context

Added 5 WTQ datasets that were failing accuracy tests with `encode()`/`decode()`.

**Model used:** `llama3.1:8b-instruct-q4_K_M` (localhost via ollama)

**Limitation:** No AZURE_OPENAI_* env vars available to run against the zon repo accuracy suite.

**Test data sources:**
- [Fixtures](https://github.com/P6rguVyrst/wick-formatter/blob/main/tests/benchmarks/conftest.py#L18)
- [Dataclasses](https://github.com/P6rguVyrst/wick-formatter/blob/main/tests/benchmarks/wtq_loader.py)

### CODE REVIEW

**Review by ronibhakta1 — CHANGES_REQUESTED** [review]
Lets keep the tests file to a different folder here to make sure we can keep a track and just make it plugin which we can plug or unplug it from the test scripts

**Comment by P6rguVyrst** [comment]
> Lets keep the tests file to a different folder here to make sure we can keep a track and just make it plugin which we can plug or unplug it from the test scripts

Thanks for getting back so quickly, I didn't have the energy mid-week to follow up, sorry. 

Added a test for testing round-trip accuracy-- test_roundtrip_benchmarks tests for all JSON files in `benchmarks/data/ `folder now, including the 5 new ones that I added.

How to run the tests that validate that https://github.com/ZON-Format/ZON/pull/10 is fixing genuine bugs with the format. Currently ZON format does not offer "lossless round-trips".

I'm surpprised that the CI isn't failing. Should probably add a workflow to CI that runs the tests before-- I can do that, but not in this PR, as that's a separate concern.

```
python -m unittest tests.unit.test_roundtrip_benchmarks -v

or...

pytest tests/unit/test_roundtrip_benchmarks.py -v
```

**Comment by ronibhakta1** [comment]
> > Lets keep the tests file to a different folder here to make sure we can keep a track and just make it plugin which we can plug or unplug it from the test scripts
> 
> Thanks for getting back so quickly, I didn't have the energy mid-week to follow up, sorry.
> 
> Added a test for testing round-trip accuracy-- test_roundtrip_benchmarks tests for all JSON files in `benchmarks/data/ `folder now, including the 5 new ones that I added.
> 
> How to run the tests that validate that #10 is fixing genuine bugs with the format. Currently ZON format does not offer "lossless round-trips".
> 
> I'm surpprised that the CI isn't failing. Should probably add a workflow to CI that runs the tests before-- I can do that, but not in this PR, as that's a separate concern.
> 
> ```
> python -m unittest tests.unit.test_roundtrip_benchmarks -v
> 
> or...
> 
> pytest tests/unit/test_roundtrip_benchmarks.py -v
> ```

Hey @P6rguVyrst Thanks for finding's. I only had round trip check's. Lossless once might have got missed out on the egde cases. 

The edge cases for roundtrip are like Spider Multiverse 
Alt Text

**We can try fixing and adding the github test's with your data sets.**

Just tell me which one should we merge first

**Comment by P6rguVyrst** [comment]
> > > Lets keep the tests file to a different folder here to make sure we can keep a track and just make it plugin which we can plug or unplug it from the test scripts
> > 
> > 
> > Thanks for getting back so quickly, I didn't have the energy mid-week to follow up, sorry.
> > Added a test for testing round-trip accuracy-- test_roundtrip_benchmarks tests for all JSON files in `benchmarks/data/ `folder now, including the 5 new ones that I added.
> > How to run the tests that validate that #10 is fixing genuine bugs with the format. Currently ZON format does not offer "lossless round-trips".
> > I'm surpprised that the CI isn't failing. Should probably add a workflow to CI that runs the tests before-- I can do that, but not in this PR, as that's a separate concern.
> > ```
> > python -m unittest tests.unit.test_roundtrip_benchmarks -v
> > 
> > or...
> > 
> > pytest tests/unit/test_roundtrip_benchmarks.py -v
> > ```
> 
> Hey @P6rguVyrst Thanks for finding's. I only had round trip check's. Lossless once might have got missed out on the egde cases.
> 
> The edge cases for roundtrip are like Spider Multiverse Alt Text [ Alt Text ](https://camo.githubusercontent.com/088c0bbd367800603238310187920196faff7c016b201e7afb800fee76e3daa4/68747470733a2f2f6d65646961332e67697068792e636f6d2f6d656469612f76312e59326c6b505463354d4749334e6a45786454523259585a7a616e45774d5852784e334e6e4d6e59335932647563334e7a4d545577596d5a6d633249314e33567a644856734e535a6c634431324d563970626e526c636d35686246396e61575a66596e6c666157516d593351395a772f6c33366b553830785066306f6a47304572672f67697068792e676966) [ ](https://camo.githubusercontent.com/088c0bbd367800603238310187920196faff7c016b201e7afb800fee76e3daa4/68747470733a2f2f6d65646961332e67697068792e636f6d2f6d656469612f76312e59326c6b505463354d4749334e6a45786454523259585a7a616e45774d5852784e334e6e4d6e59335932647563334e7a4d545577596d5a6d633249314e33567a644856734e535a6c634431324d563970626e526c636d35686246396e61575a66596e6c666157516d593351395a772f6c33366b553830785066306f6a47304572672f67697068792e676966)
> 
> **We can try fixing and adding the github test's with your data sets.**
> 
> Just tell me which one should we merge first

I was thinking of adding the failing tests first, and then the PR that's fixing them.

I don't mind if you scrap my PRs, and patch the code yourself-- just wanted to use ZON, but not have to maintain my own copy of it for accuracy 😅 

The PR history for the actual patch in my for got a bit messy, so I'd probably do a squash-and-merge for at least the 3 bug fixes-- there's a bit back and forth there of reverting a PR I accidentally merged to my main.

File: zon-format/tests/unit/test_roundtrip_benchmarks.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+        encoded = zon.encode(original)
+        decoded = zon.decode(encoded)
+
+        orig_json = json.dumps(original, sort_keys=True)
+        dec_json = json.dumps(decoded, sort_keys=True)
+
+        self.assertEqual(
+            orig_json, dec_json,
+            f"Roundtrip failed for {filepath.name}"
```
[the remark this answers is not in the document]
Comment (P6rguVyrst) [reply zon-format/tests/unit/test_roundtrip_benchmarks.py:42]: {1: "value"} isn't valid json -- we are specifically loading in json content and encoding+decoding it. 
the json.load at line 32 will fail with the example 


json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 2 column 5 (char 6)

**Comment by ronibhakta1** [comment]
> > > > Lets keep the tests file to a different folder here to make sure we can keep a track and just make it plugin which we can plug or unplug it from the test scripts
> > > 
> > > 
> > > Thanks for getting back so quickly, I didn't have the energy mid-week to follow up, sorry.
> > > Added a test for testing round-trip accuracy-- test_roundtrip_benchmarks tests for all JSON files in `benchmarks/data/ `folder now, including the 5 new ones that I added.
> > > How to run the tests that validate that #10 is fixing genuine bugs with the format. Currently ZON format does not offer "lossless round-trips".
> > > I'm surpprised that the CI isn't failing. Should probably add a workflow to CI that runs the tests before-- I can do that, but not in this PR, as that's a separate concern.
> > > ```
> > > python -m unittest tests.unit.test_roundtrip_benchmarks -v
> > > 
> > > or...
> > > 
> > > pytest tests/unit/test_roundtrip_benchmarks.py -v
> > > ```
> > 
> > 
> > Hey @P6rguVyrst Thanks for finding's. I only had round trip check's. Lossless once might have got missed out on the egde cases.
> > The edge cases for roundtrip are like Spider Multiverse Alt Text [ Alt Text ](https://camo.githubusercontent.com/088c0bbd367800603238310187920196faff7c016b201e7afb800fee76e3daa4/68747470733a2f2f6d65646961332e67697068792e636f6d2f6d656469612f76312e59326c6b505463354d4749334e6a45786454523259585a7a616e45774d5852784e334e6e4d6e59335932647563334e7a4d545577596d5a6d633249314e33567a644856734e535a6c634431324d563970626e526c636d35686246396e61575a66596e6c666157516d593351395a772f6c33366b553830785066306f6a47304572672f67697068792e676966) [ ](https://camo.githubusercontent.com/088c0bbd367800603238310187920196faff7c016b201e7afb800fee76e3daa4/68747470733a2f2f6d65646961332e67697068792e636f6d2f6d656469612f76312e59326c6b505463354d4749334e6a45786454523259585a7a616e45774d5852784e334e6e4d6e59335932647563334e7a4d545577596d5a6d633249314e33567a644856734e535a6c634431324d563970626e526c636d35686246396e61575a66596e6c666157516d593351395a772f6c33366b553830785066306f6a47304572672f67697068792e676966)
> > **We can try fixing and adding the github test's with your data sets.**
> > Just tell me which one should we merge first
> 
> I was thinking of adding the failing tests first, and then the PR that's fixing them.
> 
> I don't mind if you scrap my PRs, and patch the code yourself-- just wanted to use ZON, but not have to maintain my own copy of it for accuracy 😅
> 
> The PR history for the actual patch in my for got a bit messy, so I'd probably do a squash-and-merge for at least the 3 bug fixes-- there's a bit back and forth there of reverting a PR I accidentally merged to my main.

Hey I would love to have code fixes from your end. So I will start with the PR which adds the data set's and then you can try to add a fix to that.

**Review by ronibhakta1 — APPROVED** [review]

**Comment by ronibhakta1** [comment]
Hey @P6rguVyrst Now you can see the test being in action. Would you like to look into the logs above.

**Comment by P6rguVyrst** [comment]
> Hey @P6rguVyrst Now you can see the test being in action. Would you like to look into the logs above.

I mean my main branch will fix the failing tests.
What I could do is squash & merge my main branch to this PR to have the tests & the fix go out in a single release.

**Comment by ronibhakta1** [comment]
> > Hey @P6rguVyrst Now you can see the test being in action. Would you like to look into the logs above.
> 
> I mean my main branch will fix the failing tests. What I could do is squash & merge my main branch to this PR to have the tests & the fix go out in a single release.

That can be a better approach.

**Comment by P6rguVyrst** [comment]
> > > Hey @P6rguVyrst Now you can see the test being in action. Would you like to look into the logs above.
> > 
> > 
> > I mean my main branch will fix the failing tests. What I could do is squash & merge my main branch to this PR to have the tests & the fix go out in a single release.
> 
> That can be a better approach.

It's all in one PR now, however I don't see the workflow test runs triggering.

**Comment by ronibhakta1** [comment]
@P6rguVyrst Before we merge. do you wanna do a token test where we use token tester model to get the idea how much is getting consumed there @coderabbitai please tag the right files we have in the code base

**Comment by ronibhakta1** [comment]
> > > > Hey @P6rguVyrst Now you can see the test being in action. Would you like to look into the logs above.
> > > 
> > > 
> > > I mean my main branch will fix the failing tests. What I could do is squash & merge my main branch to this PR to have the tests & the fix go out in a single release.
> > 
> > 
> > That can be a better approach.
> 
> It's all in one PR now, however I don't see the workflow test runs triggering.

Yes,  I kept it to be triggered mnually

**Comment by P6rguVyrst** [comment]
> > > > > Hey @P6rguVyrst Now you can see the test being in action. Would you like to look into the logs above.
> > > > 
> > > > 
> > > > I mean my main branch will fix the failing tests. What I could do is squash & merge my main branch to this PR to have the tests & the fix go out in a single release.
> > > 
> > > 
> > > That can be a better approach.
> > 
> > 
> > It's all in one PR now, however I don't see the workflow test runs triggering.
> 
> Yes, I kept it to be triggered mnually

Looks like at least with this dataset the tests now pass? :) 

There was one rounding error even with the original dataset-- not sure if it's fixed all the acuracy bugs, but after fixing 3 distinct issues, it does seem more accurate than before.

Let me know if I should address fix else within this scope-- did address few CodeRabbit nitpicks yesterday..

**Comment by ronibhakta1** [comment]
> > > > > > Hey @P6rguVyrst Now you can see the test being in action. Would you like to look into the logs above.
> 
> > > > > 
> 
> > > > > 
> 
> > > > > I mean my main branch will fix the failing tests. What I could do is squash & merge my main branch to this PR to have the tests & the fix go out in a single release.
> 
> > > > 
> 
> > > > 
> 
> > > > That can be a better approach.
> 
> > > 
> 
> > > 
> 
> > > It's all in one PR now, however I don't see the workflow test runs triggering.
> 
> > 
> 
> > Yes, I kept it to be triggered mnually
> 
> 
> 
> Looks like at least with this dataset the tests now pass? :) 
> 
> 
> 
> There was one rounding error even with the original dataset-- not sure if it's fixed all the acuracy bugs, but after fixing 3 distinct issues, it does seem more accurate than before.
> 
> 
> 
> Let me know if I should address fix else within this scope-- did address few CodeRabbit nitpicks yesterday..

Yeah the test pass this time thanks for that. Speaking about code rabbit finding if the issues are minor let's fix and finish it off. Speaking about if any of the tests or datasets fails we have to figure it out cause it can lead to more issues on user's accuracy.

File: zon-format/src/zon/core/utils.py [hunk 2]
```diff
@@ -31,6 +31,33 @@ def quote_string(s: str) -> str:
...
+    if trimmed.startswith('"'):
+        try:
+            return json.loads(trimmed)
+        except json.JSONDecodeError:
+            if trimmed.endswith('"'):
+                inner = trimmed[1:-1]
+                json_str = inner.replace('""', '\\"')
+                return json.loads(f'"{json_str}"')
+
+    return trimmed
```
[the remark this answers is not in the document]
Comment (P6rguVyrst) [reply zon-format/src/zon/core/utils.py:58]: The encoder produces "f" and "true" (properly closed quotes). The malformed "foo case requires:

  1. Hand-crafted invalid ZON, OR
  2. Bug in encoder/tokenizer

The bug is not reachable with valid ZON.

Should the decoder be strict (reject malformed input) or permissive (best-effort)? 

That's a design choice, not a correctness bug.

File: zon-format/tests/unit/test_delta.py [hunk 3]
```diff
@@ -107,12 +108,41 @@ def test_deep_nesting(self):
...
+        decoded = self.decoder.decode(self.encoder.encode(data))
+
+        for original, got in zip(data, decoded):
+            # repr(float) is the shortest string that round-trips to the same
+            # double, so repr equality is equivalent to bit equality.
+            self.assertEqual(
+                repr(original['v']),
+                repr(got['v']),
+                f"float roundtrip lost precision: {original['v']!r} -> {got['v']!r}",
+            )
```
[the remark this answers is not in the document]
Comment (P6rguVyrst) [reply zon-format/tests/unit/test_delta.py:144]: The finding doesn't apply meaningfully to this test's purpose. 

>If a future regression caused decoder.decode(...) to return fewer rows than data

Row dropping isn't a realistic failure mode. If it were to happen, the benchmark roundtrip tests would fail immediately.

**Comment by P6rguVyrst** [comment]
Should I do anything else? 

I addressed CodeRabbit comments where appropriate, not going to commnet on test data which is valid json input.

**Comment by ronibhakta1** [comment]
> Should I do anything else?
> 
> I addressed CodeRabbit comments where appropriate, not going to commnet on test data which is valid json input.

Yeah, looks good to me. 
Let's get this one merged. Thank you for your contribution.

**Comment by P6rguVyrst** [comment]
> > Should I do anything else?
> > I addressed CodeRabbit comments where appropriate, not going to commnet on test data which is valid json input.
> 
> Yeah, looks good to me. Let's get this one merged. Thank you for your contribution.

I don't have the permission to merge, nor should I 🫠

**Comment by ronibhakta1** [comment]
> > > Should I do anything else?
> > > I addressed CodeRabbit comments where appropriate, not going to commnet on test data which is valid json input.
> > 
> > 
> > Yeah, looks good to me. Let's get this one merged. Thank you for your contribution.
> 
> I don't have the permission to merge, nor should I 🫠

Yeah sorry I will do that rn. I got busy yesterday to work on something.

**Comment by ronibhakta1** [comment]
@P6rguVyrst I have deployed the package. You can use it seamlessly
